geode-for-redis/src/distributedTest/java/org/apache/geode/redis/internal/commands/executor/list/RPushDUnitTest.java

```diff
@@ -0,0 +1,201 @@
+  private static HashMap<String, List<String>> keyToElementListMap;
+  private static List<Runnable> taskList;
+
+  private JedisCluster jedis;
+
+  @Before
+  public void testSetup() {
+    MemberVM locator = clusterStartUp.startLocatorVM(0);
+    clusterStartUp.startRedisVM(1, locator.getPort());
+    clusterStartUp.startRedisVM(2, locator.getPort());
+    clusterStartUp.startRedisVM(3, locator.getPort());
+    int redisServerPort = clusterStartUp.getRedisPort(1);
+    jedis = new JedisCluster(new HostAndPort(BIND_ADDRESS, redisServerPort), 10_000);
+    listHashtags = makeListHashtags();
+    keys = makeListKeys(listHashtags);
+    keyToElementListMap = new HashMap<>();
+    for (String key : keys) {
+      keyToElementListMap.put(key, makeElementList(PUSH_LIST_SIZE, key));
+    }
+
+    taskList = new ArrayList<>();
+    taskList.add(() -> rpushPerformAndVerify(keys.get(0), keyToElementListMap.get(keys.get(0)),
+        runningCount));
+    taskList.add(() -> rpushPerformAndVerify(keys.get(0), keyToElementListMap.get(keys.get(0)),
+        runningCount));
+    taskList.add(() -> rpushPerformAndVerify(keys.get(1), keyToElementListMap.get(keys.get(1)),
+        runningCount));
+    taskList.add(() -> rpushPerformAndVerify(keys.get(1), keyToElementListMap.get(keys.get(1)),
+        runningCount));
+    taskList.add(() -> rpushPerformAndVerify(keys.get(2), keyToElementListMap.get(keys.get(2)),
+        runningCount));
+    taskList.add(() -> rpushPerformAndVerify(keys.get(2), keyToElementListMap.get(keys.get(2)),
+        runningCount));
+    taskList.add(() -> verifyListLengthCondition(keys.get(0), runningCount));
+    taskList.add(() -> verifyListLengthCondition(keys.get(1), runningCount));
+    taskList.add(() -> verifyListLengthCondition(keys.get(2), runningCount));
+  }
+
+  @After
+  public void tearDown() {
+    jedis.close();
+  }
+
+  @Test
+  public void givenBucketsMovedDuringRPush_elementsAreAddedAtomically()
+      throws ExecutionException, InterruptedException {
+
+    List<Future<Void>> futureList = new ArrayList<>();
+    for (Runnable task : taskList) {
+      futureList.add(executor.runAsync(task));
+    }
+
+    for (int i = 0; i < 10 && runningCount.get() > 0; i++) {
+      clusterStartUp.moveBucketForKey(listHashtags.get(i % listHashtags.size()));
+      Thread.sleep(500);
+    }
+
+    for (Future<Void> future : futureList) {
+      future.get();
+    }
+
+    for (String key : keys) {
+      long length = jedis.llen(key);
+      assertThat(length).isGreaterThanOrEqualTo(MINIMUM_ITERATIONS * 2 * PUSH_LIST_SIZE);
+      validateListContents(key, keyToElementListMap);
+    }
+  }
+
+  @Test
+  public void shouldNotLoseData_givenPrimaryServerCrashesDuringOperations()
+      throws ExecutionException, InterruptedException {
+    List<Future<Void>> futureList = new ArrayList<>();
+    for (Runnable task : taskList) {
+      futureList.add(executor.runAsync(task));
+    }
+
+    Thread.sleep(200);
```

Comment:
Rather than a `Thread.sleep()` here, it would probably be better to use an `await()` until all the lists have some minimum size (since I assume that's what we're actually waiting for).